subprojects/execution/src/main/java/org/gradle/internal/execution/steps/ExecuteWorkBuildOperationFiringStep.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.execution.steps;
+
+import org.gradle.internal.Try;
+import org.gradle.internal.execution.ExecutionEngine;
+import org.gradle.internal.execution.UnitOfWork;
+import org.gradle.internal.operations.BuildOperationDescriptor;
+import org.gradle.internal.operations.BuildOperationExecutor;
+import org.gradle.operations.execution.ExecuteWorkBuildOperationType;
+
+import javax.annotation.Nullable;
+
+/**
+ * A step that executes a unit of work and wraps it into a {@link ExecuteWorkBuildOperationType} build operation.
+ */
+public class ExecuteWorkBuildOperationFiringStep<C extends IdentityContext, R extends CachingResult> extends BuildOperationStep<C, R> implements Step<C, R> {
+
+    private final Step<? super C, R> delegate;
+
+    public ExecuteWorkBuildOperationFiringStep(BuildOperationExecutor buildOperationExecutor, Step<C, R> delegate) {
+        super(buildOperationExecutor);
+        this.delegate = delegate;
+    }
+
+    @Override
+    public R execute(UnitOfWork work, C context) {
+        return operation(operationContext -> {
+                R result = delegate.execute(work, context);
+                ExecuteWorkBuildOperationType.Result operationResult = new ExecuteWorkResult(result.getExecution());
+                operationContext.setResult(operationResult);
+                return result;
+            },
+            BuildOperationDescriptor
+                .displayName("Execute unit of work")
+                .details(new ExecuteWorkDetails(work)));
+    }
+
+    private static class ExecuteWorkDetails implements ExecuteWorkBuildOperationType.Details {
+
+        private final UnitOfWork work;
+
+        public ExecuteWorkDetails(UnitOfWork work) {
+            this.work = work;
+        }
+
+        @Nullable
+        @Override
+        public String getWorkType() {
+            return work.getBuildOperationWorkType().orElse(null);
+        }
```

Comment:
❓ I don't think this is the right way to handle the type not being present. Instead, I'd skip emitting the build operation completely if `getBuildOperationWorkType()` returned empty. WDYT?

Reply:
I don't think we want to have different structure of build operations depending on some fields of the details.

I guess it's the primary reason why @wolfs wanted to keep this non-nullable.
We can also consider returning a fixed value here like `UNKNOWN`.

Reply:
After chatting with @wolfs, always emitting the build op and returning `null` seems fine.